tuning/trainercontroller/callback.py

```diff
@@ -0,0 +1,250 @@
+                            # Initialize the controller-metric instance
+                            if 'arguments' in cm:
+                                obj = class_type(cm['name'], **cm['arguments'])
+                            else:
+                                obj = class_type(cm['name'])
+                        except Exception as e:
+                            logger.fatal(e)
+                        assert (obj.validate(training_args)), 'Controller metric class [%s] cannot be computed because the training args do not support it' % (cm['handler-class'])
+                        controller_metric_objs.append(obj)
+                    self.__controllers[name] = controller_metric_objs
+        else:
+            raise ValueError("Trainer controller configuration [%s] does NOT exist" % trainer_controller_args.trainer_controller_config_file)
+
+    def __validate_config(self, config):
+        assert 'controllers' in config and len(config['controllers']) > 0, "List of controllers missing in config"
+        for c in config['controllers']:
+            assert 'name' in c, f"Controller should have a name"
+            assert ('triggers' in c) and (len(c['triggers']) > 0), f"Triggers not specified for controller {c['name']}"
+            assert ('rule' in c) and (len(c['rule']) > 0), f"Rule not specified for controller {c['name']}"
+            try:
+                compile(c['rule'], '<stdin>', 'eval')
+            except Exception as e:
+                raise SyntaxError(f"Rule for controller {c['name']} has this error {e}")
+            assert ('controller-metrics' in c) and (len(c['controller-metrics']) > 0), f"List of controller metrics missing for controller {c['name']}"
+            for cm in c['controller-metrics']:
+                assert ('name' in cm), f"Controller metric should have a name"
+                assert ('handler-class' in cm), f"Handler class not specified for controller metric {cm['name']}"
+            assert ('control-operations' in c) and (len(c['control-operations']) > 0), f"Control operations not specified for controller {c['name']}"
+            for k, v in c['control-operations'].items():
+                try:
+                    getattr(TrainerControl(), k)
+                except Exception as e:
+                    raise AttributeError(f"Control operation {k} is not invalid for controller {c['name']}")
+
+                assert (isinstance(v, bool) or isinstance(v, int)),  f"Control operation {k} is assigned invalid value for controller {c['name']}"
+
+
+    def __apply_control(self, cb, control):
+        """Given a controller-block, applies the control operation to the training loop.
+
+        Args:
+            cb: Controller block dictionary
+            control: TrainerControl object
+
+        Returns:
+            None.
+        """
+        for k, v in cb['control-operations'].items():
+            setattr(control, k, v)
```

Comment:
We don't need a function for it I guess.

Reply:
done.